x-pack/plugins/enterprise_search/public/applications/shared/use_local_storage/use_local_storage.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useState, useEffect } from 'react';
+
+export const useLocalStorage = <T>(key: string, defaultValue: T) => {
+  const [item, setItem] = useState<T>(getFromStorage());
+
+  function getFromStorage() {
+    const storedItem = window.localStorage.getItem(key);
+    let toStore: T = defaultValue;
+
+    if (storedItem !== null) {
+      try {
+        toStore = JSON.parse(storedItem) as T;
+      } catch (err) {
+        window.localStorage.removeItem(key);
+      }
+    }
+
+    return toStore;
+  }
+
+  const updateFromStorage = () => {
+    const storedItem = getFromStorage();
+    setItem(storedItem);
+  };
+
+  const saveToStorage = (value: T) => {
+    window.localStorage.setItem(key, JSON.stringify(value));
+    updateFromStorage();
+  };
+
+  useEffect(() => {
+    const handler = (event: StorageEvent) => {
+      if (event.key === key) {
+        updateFromStorage();
+      }
+    };
+    window.addEventListener('storage', handler);
+    return () => window.removeEventListener('storage', handler);
+  }, []);
+
+  return [item, saveToStorage] as const;
```

Comment:
whoa, this typescript cast is tripping me out. how does this work exactly? or should we be more explicit about the types being returned?

```suggestion
  return [item, saveToStorage] as [T, Function];
```

Reply:
3b0aae058f05ab63e392beddab80221b21b9eedf

Reply:
3b0aae058f05ab63e392beddab80221b21b9eedf

Reply:
Nice!! Even better 👏